### PR TITLE
RenderComponent - new component to replace ModelComponent

### PR DESCRIPTION
**Pull request by mvaligursky**
Related to #2063

Implementation of new RenderComponent, which is functionally similar to ModelComponent, and main reason for its development is to expose Entity hierarchy of GLB files to engine and eventually Editor. Glb file will be loaded as a hierarchy of entities, and RenderComponents would be added on any of these Entities to represent renderable meshes.

This is implemented as @private functionality for the moment, and the API is subject to change.

New API:
- ContainerResource.instantiateModelEntity - creates and entity and adds ModelComponent for model stores as ContainerResource.model.  

- ContainerResource.instantiateRenderEntity - this is the new functionality, and it builds hierarchy of Entities and adds RenderComponents for relevant entities representing meshes assigned to them.

RenderComponent API (similar to ModelComponent API)
- **.meshInstances** - array of meshesInstances this component renders
- **.hide, .show** - allows attached instances to change visibility
- **.generateWireframe** - generates index buffer used to wireframe rendering
- **.castShadows**
- **.receiveShadows**
- **.castShadowsLightmap**
- **.lightmapped, .lightmapSizeMultiplier**
- **.isStatic**
- **.batchGroupId** Assign meshes to a specific batch group
- **.material** The material that will be used to render the meshes (not used on models of type 'asset').

Example:

```
entity.addComponent('render', {
                type: "sphere",
                material: someMaterial,
                castShadows: false,
                castShadowsLightmap: true,
                lightmapped: true
            });
```
 
Other changes:
- Anim system supports Entity hierarchies and RenderComponents (for animation of Entity transforms, and RenderComponent morph weights)
- Batching - supports batching of meshes rendered using RenderComponent
- lightmapping - supports lightmapping of meshes rendered using RenderComponent

### CODE REVIEW

File: src/scene/lightmapper.js [hunk 1]
```diff
@@ -446,10 +489,10 @@ Object.assign(Lightmapper.prototype, {
...
         var meshes;
         for (node = 0; node < allNodes.length; node++) {
-            origCastShadows[node] = allNodes[node].model.castShadows;
-            allNodes[node].model.castShadows = allNodes[node].model.castShadowsLightmap;
-            if (allNodes[node].model.castShadowsLightmap) {
-                meshes = allNodes[node].model.meshInstances;
+            origCastShadows[node] = renderOrModel(allNodes[node]).castShadows;
+            renderOrModel(allNodes[node]).castShadows = renderOrModel(allNodes[node]).castShadowsLightmap;
+            if (renderOrModel(allNodes[node]).castShadowsLightmap) {
+                meshes = renderOrModelMeshInstances(allNodes[node]);
```
Comment (yaustar) [inline src/scene/lightmapper.js:495]: Style: We can cache `renderOrModel()` over calling it multiple times here.

Comment (mvaligursky) [reply src/scene/lightmapper.js]: done

File: src/anim/anim.js [hunk 2]
```diff
@@ -640,6 +640,23 @@ function DefaultAnimBinder(graph) {
     };
     flatten(graph);
 
+    var findMeshInstances = function (node) {
+        var object = node;
+        while (object && object.constructor !== Entity) {
```
Comment (vkalpias) [inline src/anim/anim.js:645]: I think the preferred way to check this is ``object instanceof Entity`

File: src/framework/components/system.js [hunk 3]
```diff
@@ -21,6 +23,9 @@ function ComponentSystem(app) {
     // The store where all pc.ComponentData objects are kept
     this.store = {};
     this.schema = [];
+
+    // cached mesh primitives
+    this.primitiveMap = new Map();
```
Comment (vkalpias) [inline src/framework/components/system.js:28]: This feels like it should be somewhere graphics related and not in the ComponentSystem?

File: src/framework/components/render/system.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+            castShadows: entity.render.castShadows,
+            receiveShadows: entity.render.receiveShadows,
+            castShadowsLightmap: entity.render.castShadowsLightmap,
+            lightmapped: entity.render.lightmapped,
+            lightmapSizeMultiplier: entity.render.lightmapSizeMultiplier,
+            isStatic: entity.render.isStatic,
+            enabled: entity.render.enabled,
+            layers: entity.render.layers,
+            batchGroupId: entity.render.batchGroupId,
+            mapping: extend({}, entity.render.mapping)
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/framework/components/render/system.js]: this is copied from model component
but as this line is not needed (for now at least), I removed it.

File: src/anim/anim.js (hunk 2 above)
Comment (mvaligursky) [reply src/anim/anim.js:645]: I'll fix. @ellthompson - I grabbed this line from anim.js, maybe improve it there as well.

Comment (mvaligursky) [reply src/anim/anim.js:645]: fair point, I only refactored existing code, this was there. I'll change it.

Comment (mvaligursky) [reply src/anim/anim.js]: done

File: src/framework/components/system.js [hunk 5]
```diff
@@ -281,6 +286,57 @@ Object.assign(ComponentSystem.prototype, {
 
     destroy: function () {
         this.off();
+    },
+
+    // returns Primive data, used by ModelComponent and RenderComponent
+    getPrimitive: function (type) {
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/framework/components/system.js]: done, cheers!

File: src/framework/components/system.js (hunk 3 above)
Comment (mvaligursky) [reply src/framework/components/system.js]: done, thanks!

File: src/framework/components/render/component.js [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+        if (this._castShadows !== value) {
+
+            var i, layer, mi = this._meshInstances;
+
+            if (mi) {
+                var layers = this.layers;
+                var scene = this.system.app.scene;
+                if (this._castShadows && !value) {
+                    for (i = 0; i < layers.length; i++) {
+                        layer = this.system.app.scene.layers.getLayerById(this.layers[i]);
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/framework/components/render/component.js:365]: fixing, thanks.

File: src/framework/components/render/component.js [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+
+    set: function (value) {
+        if (this._receiveShadows !== value) {
+
+            this._receiveShadows = value;
+
+            var mi = this._meshInstances;
+            if (mi) {
+                for (var i = 0; i < mi.length; i++) {
+                    meshInstances[i].receiveShadow = value;
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/framework/components/render/component.js:404]: fixing, thanks!

File: src/framework/components/render/component.js [hunk 8]
```diff
@@ -0,0 +1,526 @@
...
+
+            var meshInstances = this._meshInstances;
+            for (var i = 0; i < meshInstances.length; i++) {
+                meshInstances[i].castShadow = this._castShadows;
+                meshInstances[i].receiveShadow = this._receiveShadows;
+                meshInstances[i].isStatic = this._isStatic;
+            }
+
+            // update meshInstances
+            this.lightmapped = this._lightmapped;
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/framework/components/render/component.js]: done too, thanks, well spotted.

**Comment by tonetechnician** [comment]
Hey @mvaligursky 

Just to confirm that I've been having great success with the instantiateRenderEntity and instantiateModelEntity functions my side. I've been able to load and apply individual scripts to each of the assets in my scene. Great work!

**Pull request closed without merging**